orquesta/tests/unit/conducting/test_workflow_conductor_rerun.py

```diff
@@ -0,0 +1,425 @@
+  task2:
+    action: core.noop
+
+  # branch 2
+  task3:
+    action: core.noop
+    next:
+      - when: <% succeeded() %>
+        publish:
+          - var2: 123
+        do: task4
+  task4:
+    action: core.noop
+
+output:
+  - var1: <% ctx().var1 %>
+  - var2: <% ctx().var2 %>
+"""
+
+JOIN_WF_DEF = """
+version: 1.0
+
+description: A basic branching workflow.
+
+tasks:
+  # branch 1
+  task1:
+    action: core.noop
+    next:
+      - when: <% succeeded() %>
+        publish:
+          - var1: 'xyz'
+        do: task3
+
+  # branch 2
+  task2:
+    action: core.noop
+    next:
+      - when: <% succeeded() %>
+        publish:
+          - var2: 123
+        do: task3
+
+  # joining branch
+  task3:
+    join: all
+    action: core.noop
+    next:
+      - when: <% succeeded() %>
+        publish:
+          - var3: True
+        do: task4
+  task4:
+    action: core.noop
+"""
+
+
+class WorkflowConductorRerunTest(test_base.WorkflowConductorTest):
+
+    def pre_rerun_workflow(self, wf_def):
+        spec = native_specs.WorkflowSpec(wf_def)
+        self.assertDictEqual(spec.inspect(), {})
+
+        # Run the workflow.
+        conductor = conducting.WorkflowConductor(spec)
+        conductor.request_workflow_status(statuses.RUNNING)
+
+        return conductor
+
+    def validate_request_workflow_rerun(self, conductor, rerun_tasks, number_tasks):
+        options = dict()
+        options['tasks'] = rerun_tasks
+        conductor.request_workflow_rerun(options)
+
+        # Assert the workflow status.
+        self.assertEqual(conductor.get_workflow_status(), statuses.RESUMING)
+
+        # Assert the number of tasks, staged tasks
+        tasks = conductor.workflow_state.get_tasks()
+        self.assertEqual(len(tasks), number_tasks + len(rerun_tasks))
+
+        staged_tasks = conductor.workflow_state.get_staged_tasks()
+        self.assertEqual(len(staged_tasks), len(rerun_tasks))
+
+        for i in range(len(staged_tasks)):
+            self.assertEqual(staged_tasks[i]['id'], rerun_tasks[i])
+            self.assertEqual(staged_tasks[i]['ready'], True)
+
+        # Assert the workflow status.
+        self.assertEqual(conductor.get_workflow_status(), statuses.RESUMING)
+
+    def test_parallel_rerun_with_first_task(self):
+        expected_output = {
+            'var1': 'xyz',
+            'var2': 123
+        }
+
+        # 1. Prepare rerun workflow
+        conductor = self.pre_rerun_workflow(PARALLEL_WF_DEF)
+
+        # 2. Fail task1
+        # Conduct tasks.
+        self.forward_task_statuses(conductor, 'task1', [statuses.RUNNING])
+        self.forward_task_statuses(conductor, 'task3', [statuses.RUNNING])
+
+        self.forward_task_statuses(conductor, 'task1', [statuses.FAILED])
+        self.forward_task_statuses(conductor, 'task3', [statuses.SUCCEEDED])
+
+        self.forward_task_statuses(conductor, 'task4', [statuses.RUNNING])
+        self.forward_task_statuses(conductor, 'task4', [statuses.SUCCEEDED])
```

Comment:
Can we move running of task4 until after rerun?  I want to make sure unfinished tasks from parallel branch will be picked up.

Reply:
You have not addressed this yet.